Enable strict mypy checking

A follow-on from #326. Using the `--strict` flag with `mypy` produces another 98 errors that will need to be squashed.

Note that currently I get 137 errors for this:
```
mitiq/mitiq_pyquil/pyquil_utils.py:33: error: Function is missing a type annotation
mitiq/mitiq_pyquil/pyquil_utils.py:105: error: Returning Any from function declared to return "float"
mitiq/mitiq_pyquil/pyquil_utils.py:121: error: Returning Any from function declared to return "float"
mitiq/mitiq_pyquil/pyquil_utils.py:169: error: Function is missing a type annotation
mitiq/mitiq_pyquil/compiler.py:219: error: Returning Any from function declared to return "bool"
mitiq/utils.py:120: error: Returning Any from function declared to return "bool"
mitiq/utils.py:125: error: Implicit generic "Any". Use "typing.Dict" and specify generic parameters
mitiq/benchmarks/utils.py:47: error: Returning Any from function declared to return "float"
mitiq/benchmarks/randomized_benchmarking.py:62: error: unused 'type: ignore' comment
mitiq/benchmarks/randomized_benchmarking.py:73: error: unused 'type: ignore' comment
mitiq/benchmarks/randomized_benchmarking.py:77: error: unused 'type: ignore' comment
mitiq/benchmarks/randomized_benchmarking.py:78: error: unused 'type: ignore' comment
mitiq/tests/test_utils.py:32: error: Function is missing a type annotation
mitiq/tests/test_utils.py:34: error: Module has no attribute "H"
mitiq/tests/test_utils.py:35: error: Module has no attribute "H"
mitiq/tests/test_utils.py:41: error: Function is missing a type annotation
mitiq/tests/test_utils.py:47: error: Module has no attribute "H"
mitiq/tests/test_utils.py:48: error: Module has no attribute "H"
mitiq/tests/test_utils.py:53: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:53: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:57: error: Module has no attribute "H"
mitiq/tests/test_utils.py:58: error: Module has no attribute "H"
mitiq/tests/test_utils.py:63: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:63: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:67: error: Module has no attribute "H"
mitiq/tests/test_utils.py:68: error: Module has no attribute "H"
mitiq/tests/test_utils.py:73: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:73: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:77: error: Module has no attribute "H"
mitiq/tests/test_utils.py:78: error: Module has no attribute "H"
mitiq/tests/test_utils.py:83: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:83: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:87: error: Module has no attribute "H"
mitiq/tests/test_utils.py:88: error: Module has no attribute "H"
mitiq/tests/test_utils.py:93: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:93: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:97: error: Module has no attribute "H"
mitiq/tests/test_utils.py:98: error: Module has no attribute "H"
mitiq/tests/test_utils.py:103: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:103: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:107: error: Module has no attribute "H"
mitiq/tests/test_utils.py:108: error: Module has no attribute "H"
mitiq/tests/test_utils.py:113: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:113: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:114: error: Module has no attribute "random_circuit"
mitiq/tests/test_utils.py:130: error: Function is missing a type annotation
mitiq/tests/test_utils.py:133: error: Module has no attribute "random_circuit"
mitiq/tests/test_utils.py:149: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:149: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:150: error: Module has no attribute "random_circuit"
mitiq/tests/test_utils.py:153: error: Module has no attribute "random_circuit"
mitiq/tests/test_utils.py:171: error: Function is missing a type annotation
mitiq/tests/test_utils.py:174: error: Module has no attribute "random_circuit"
mitiq/tests/test_utils.py:177: error: Module has no attribute "random_circuit"
mitiq/tests/test_utils.py:196: error: Function is missing a type annotation
mitiq/tests/test_utils.py:204: error: Function is missing a type annotation
mitiq/tests/test_utils.py:209: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:209: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:240: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:240: note: Use "-> None" if function does not return a value
mitiq/tests/test_utils.py:259: error: Function is missing a return type annotation
mitiq/tests/test_utils.py:259: note: Use "-> None" if function does not return a value
mitiq/mitiq_pyquil/conversions.py:22: error: Module 'cirq.contrib.quil_import' has no attribute 'circuit_from_quil'
mitiq/mitiq_pyquil/conversions.py:40: error: Returning Any from function declared to return "str"
mitiq/mitiq_pyquil/conversions.py:42: error: Returning Any from function declared to return "str"
mitiq/mitiq_qiskit/conversions.py:20: error: Module 'cirq.contrib.qasm_import' has no attribute 'circuit_from_qasm'
mitiq/mitiq_qiskit/conversions.py:39: error: Returning Any from function declared to return "str"
mitiq/zne/scaling.py:54: error: Module has no attribute "H"
mitiq/zne/scaling.py:55: error: Module has no attribute "X"
mitiq/zne/scaling.py:56: error: Module has no attribute "Y"
mitiq/zne/scaling.py:57: error: Module has no attribute "Z"
mitiq/zne/scaling.py:58: error: Module has no attribute "T"
mitiq/zne/scaling.py:59: error: Module has no attribute "I"
mitiq/zne/scaling.py:60: error: Module has no attribute "CNOT"; maybe "CCNOT"?
mitiq/zne/scaling.py:61: error: Module has no attribute "CZ"; maybe "CCZ"?
mitiq/zne/scaling.py:62: error: Module has no attribute "TOFFOLI"
mitiq/zne/scaling.py:67: error: Name 'ops.Operation' is not defined
mitiq/zne/scaling.py:76: error: Name 'ops.Operation' is not defined
mitiq/zne/scaling.py:91: error: Name 'ops.Operation' is not defined
mitiq/zne/scaling.py:99: error: Name 'ops.Operation' is not defined
mitiq/zne/scaling.py:173: error: Function is missing a type annotation
mitiq/zne/scaling.py:212: error: Function is missing a type annotation
mitiq/zne/scaling.py:232: error: Missing type parameters for generic type "Callable"
mitiq/zne/scaling.py:236: error: Function is missing a type annotation for one or more arguments
mitiq/zne/scaling.py:343: error: Name 'ops.Operation' is not defined
mitiq/zne/scaling.py:343: error: Function is missing a return type annotation
mitiq/zne/scaling.py:349: error: Name 'ops.Operation' is not defined
mitiq/zne/scaling.py:360: error: Returning Any from function declared to return "float"
mitiq/zne/scaling.py:374: error: Returning Any from function declared to return "float"
mitiq/zne/scaling.py:403: error: Function is missing a type annotation for one or more arguments
mitiq/zne/scaling.py:524: error: Function is missing a type annotation for one or more arguments
mitiq/zne/scaling.py:602: error: Implicit generic "Any". Use "typing.Dict" and specify generic parameters
mitiq/zne/scaling.py:603: error: Implicit generic "Any". Use "typing.Dict" and specify generic parameters
mitiq/zne/scaling.py:638: error: Function is missing a type annotation for one or more arguments
mitiq/zne/scaling.py:796: error: Function is missing a type annotation for one or more arguments
mitiq/zne/scaling.py:863: error: Function is missing a type annotation for one or more arguments
mitiq/mitiq_qiskit/qiskit_utils.py:87: error: Returning Any from function declared to return "float"
mitiq/mitiq_qiskit/qiskit_utils.py:156: error: Returning Any from function declared to return "float"
mitiq/zne/parameter.py:65: error: Call to untyped function "_get_base_gate" in typed context
mitiq/zne/parameter.py:79: error: Function is missing a type annotation
mitiq/_about.py:24: error: Module 'cirq' has no attribute '__version__'; maybe "_version"?
mitiq/_about.py:37: error: Function is missing a return type annotation
mitiq/_about.py:37: note: Use "-> None" if function does not return a value
mitiq/_about.py:66: error: Module has no attribute "__version__"
mitiq/_about.py:71: error: Call to untyped function "about" in typed context
mitiq/zne/zne.py:28: error: Incompatible default for argument "factory" (default has type "None", argument has type "Factory")
mitiq/zne/zne.py:67: error: Incompatible default for argument "factory" (default has type "None", argument has type "Factory")
mitiq/zne/zne.py:97: error: Incompatible default for argument "factory" (default has type "None", argument has type "Factory")
mitiq/zne/inference.py:26: error: Module 'mitiq' has no attribute 'QPROGRAM'
mitiq/zne/inference.py:178: error: Implicit generic "Any". Use "typing.Dict" and specify generic parameters
mitiq/zne/inference.py:287: error: Function is missing a type annotation for one or more arguments
mitiq/zne/inference.py:294: error: Returning Any from function declared to return "float"
mitiq/zne/inference.py:298: error: Function is missing a type annotation
mitiq/zne/inference.py:391: error: Function is missing a type annotation
mitiq/zne/inference.py:392: error: Call to untyped function "__eq__" in typed context
mitiq/zne/inference.py:455: error: Function is missing a type annotation
mitiq/zne/inference.py:456: error: Call to untyped function "__eq__" in typed context
mitiq/zne/inference.py:586: error: Function is missing a type annotation
mitiq/zne/inference.py:596: error: Call to untyped function "__eq__" in typed context
mitiq/zne/inference.py:600: error: Call to untyped function "__eq__" in typed context
mitiq/zne/inference.py:680: error: Implicit generic "Any". Use "typing.Dict" and specify generic parameters
mitiq/zne/inference.py:751: error: Function is missing a return type annotation
mitiq/zne/inference.py:757: error: Function is missing a return type annotation
mitiq/zne/inference.py:806: error: Function is missing a type annotation
mitiq/zne/inference.py:808: error: Call to untyped function "__eq__" in typed context
mitiq/zne/inference.py:946: error: Function is missing a type annotation for one or more arguments
mitiq/zne/inference.py:947: error: Returning Any from function declared to return "bool"
mitiq/zne/inference.py:948: error: Call to untyped function "__eq__" in typed context
mitiq/__init__.py:19: error: Module 'mitiq.zne' has no attribute 'execute_with_zne'
mitiq/__init__.py:19: error: Module 'mitiq.zne' has no attribute 'mitigate_executor'
mitiq/__init__.py:19: error: Module 'mitiq.zne' has no attribute 'zne_decorator'
mitiq/benchmarks/random_circuits.py:20: error: Module 'cirq.testing' has no attribute 'random_circuit'
mitiq/benchmarks/random_circuits.py:64: error: Incompatible default for argument "fac" (default has type "None", argument has type "Factory")
mitiq/benchmarks/random_circuits.py:124: error: Function is missing a return type annotation
mitiq/benchmarks/maxcut.py:56: error: Missing type parameters for generic type "Callable"
mitiq/benchmarks/maxcut.py:56: error: Incompatible default for argument "scale_noise" (default has type "None", argument has type "Callable[..., Any]")
mitiq/benchmarks/maxcut.py:57: error: Incompatible default for argument "factory" (default has type "None", argument has type "Factory")
mitiq/benchmarks/maxcut.py:84: error: Function is missing a type annotation
mitiq/benchmarks/maxcut.py:87: error: Function is missing a type annotation
mitiq/benchmarks/maxcut.py:92: error: Function is missing a type annotation
mitiq/benchmarks/maxcut.py:97: error: Call to untyped function "cost_step" in typed context
mitiq/benchmarks/maxcut.py:97: error: Call to untyped function "mix_step" in typed context
mitiq/benchmarks/maxcut.py:113: error: Function is missing a type annotation
mitiq/benchmarks/maxcut.py:114: error: Call to untyped function "qaoa_ansatz" in typed context
mitiq/benchmarks/maxcut.py:132: error: Missing type parameters for generic type "Callable"
mitiq/benchmarks/maxcut.py:132: error: Incompatible default for argument "scale_noise" (default has type "None", argument has type "Callable[..., Any]")
mitiq/benchmarks/maxcut.py:133: error: Incompatible default for argument "factory" (default has type "None", argument has type "Factory")
mitiq/benchmarks/maxcut.py:135: error: Missing type parameters for generic type "List"
mitiq/benchmarks/maxcut.py:168: error: Function is missing a type annotation for one or more arguments
Found 137 errors in 17 files (checked 25 source files)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,7 @@
 ### All Changes
 
 - Add arxiv badge to readme header (@nathanshammah, gh-376).
+- Make mypy checking stricter (@karalekas, gh-380).
 
 ## Version 0.1.0 (September 2nd, 2020)
 

--- a/mitiq/_about.py
+++ b/mitiq/_about.py
@@ -34,7 +34,7 @@ MITIQ_INSTALL_PATH = os.path.dirname(MITIQ_SOURCE_FILE)
 PYTHON_VERSION = sys.version_info[0:3]
 
 
-def about():
+def about() -> None:
     """
     About box for Mitiq. Gives version numbers for
     Mitiq, NumPy, SciPy, Cirq, PyQuil, Qiskit.

--- a/mitiq/benchmarks/maxcut.py
+++ b/mitiq/benchmarks/maxcut.py
@@ -130,7 +130,7 @@ def run_maxcut(
     graph: List[Tuple[int, int]],
     x0: np.ndarray,
     noise: float = 0,
-    scale_noise:  Optional[Callable[[QPROGRAM, float], QPROGRAM]] = None,
+    scale_noise: Optional[Callable[[QPROGRAM, float], QPROGRAM]] = None,
     factory: Optional[Factory] = None,
     verbose: bool = True,
 ) -> Tuple[float, np.ndarray, List[float]]:

--- a/mitiq/benchmarks/maxcut.py
+++ b/mitiq/benchmarks/maxcut.py
@@ -16,7 +16,7 @@
 """This module contains methods for benchmarking mitiq error extrapolation
 against a standard QAOA for MAXCUT.
 """
-from typing import List, Tuple, Callable
+from typing import Callable, List, Optional, Tuple
 import numpy as np
 
 from cirq import Circuit, NamedQubit, X, ZZ, H, DensityMatrixSimulator
@@ -24,7 +24,7 @@ from cirq import identity_each as id
 
 from scipy.optimize import minimize
 
-from mitiq import execute_with_zne
+from mitiq import execute_with_zne, QPROGRAM
 from mitiq.zne.inference import Factory
 from mitiq.benchmarks.utils import noisy_simulation
 
@@ -53,8 +53,8 @@ def make_noisy_backend(
 def make_maxcut(
     graph: List[Tuple[int, int]],
     noise: float = 0,
-    scale_noise: Callable = None,
-    factory: Factory = None,
+    scale_noise: Optional[Callable[[QPROGRAM, float], QPROGRAM]] = None,
+    factory: Optional[Factory] = None,
 ) -> Tuple[
     Callable[[np.ndarray], float], Callable[[np.ndarray], Circuit], np.ndarray
 ]:
@@ -81,15 +81,15 @@ def make_maxcut(
     # one qubit per node
     qreg = [NamedQubit(str(nn)) for nn in nodes]
 
-    def cost_step(beta):
+    def cost_step(beta: float) -> Circuit:
         return Circuit(ZZ(qreg[u], qreg[v]) ** (beta) for u, v in graph)
 
-    def mix_step(gamma):
+    def mix_step(gamma: float) -> Circuit:
         return Circuit(X(qq) ** gamma for qq in qreg)
 
     init_state_prog = Circuit(H.on_each(qreg))
 
-    def qaoa_ansatz(params):
+    def qaoa_ansatz(params: np.ndarray) -> Circuit:
         half = int(len(params) / 2)
         betas, gammas = params[:half], params[half:]
         qaoa_steps = sum(
@@ -110,11 +110,12 @@ def make_maxcut(
     noisy_backend = make_noisy_backend(noise, cost_mat)
 
     # must have this function signature to work with scipy minimize
-    def qaoa_cost(params):
+    def qaoa_cost(params: np.ndarray) -> float:
         qaoa_prog = qaoa_ansatz(params)
         if scale_noise is None and factory is None:
             return noisy_backend(qaoa_prog)
         else:
+            assert scale_noise is not None
             return execute_with_zne(
                 qaoa_prog,
                 executor=noisy_backend,
@@ -129,10 +130,10 @@ def run_maxcut(
     graph: List[Tuple[int, int]],
     x0: np.ndarray,
     noise: float = 0,
-    scale_noise: Callable = None,
-    factory: Factory = None,
+    scale_noise:  Optional[Callable[[QPROGRAM, float], QPROGRAM]] = None,
+    factory: Optional[Factory] = None,
     verbose: bool = True,
-) -> Tuple[float, np.ndarray, List]:
+) -> Tuple[float, np.ndarray, List[float]]:
     """Solves MAXCUT using QAOA on a cirq wavefunction simulator using a
        Nelder-Mead optimizer.
 
@@ -143,6 +144,7 @@ def run_maxcut(
         noise: The level of depolarizing noise.
         scale_noise: The noise scaling method for ZNE.
         factory: The factory to use for ZNE.
+        verbose: An option to pass to minimize.
 
     Returns:
         A triple of the minimum cost, the values of beta and gamma that
@@ -165,7 +167,7 @@ def run_maxcut(
     # store the optimization trajectories
     traj = []
 
-    def callback(xk) -> bool:
+    def callback(xk: np.ndarray) -> bool:
         traj.append(qaoa_cost(xk))
         return True
 

--- a/mitiq/benchmarks/random_circuits.py
+++ b/mitiq/benchmarks/random_circuits.py
@@ -61,7 +61,7 @@ def rand_circuit_zne(
     depth: int,
     trials: int,
     noise: float,
-    fac: Factory = None,
+    fac: Optional[Factory] = None,
     scale_noise: Callable[[QPROGRAM, float], QPROGRAM] = fold_gates_at_random,
     op_density: float = 0.99,
     silent: bool = True,
@@ -121,7 +121,7 @@ def rand_circuit_zne(
         assert np.isreal(exact)
 
         # create the simulation type
-        def obs_sim(circ: Circuit):
+        def obs_sim(circ: Circuit) -> float:
             # we only want the expectation value not the variance
             # this is why we return [0]
             return noisy_simulation(circ, noise, obs)

--- a/mitiq/conversions.py
+++ b/mitiq/conversions.py
@@ -118,7 +118,9 @@ def convert_from_mitiq(circuit: Circuit, conversion_type: str) -> QPROGRAM:
     return converted_circuit
 
 
-def converter(noise_scaling_function: Callable[..., Any]) -> Callable[..., Any]:
+def converter(
+    noise_scaling_function: Callable[..., Any]
+) -> Callable[..., Any]:
     """Decorator for handling conversions.
 
     Args:
@@ -127,7 +129,9 @@ def converter(noise_scaling_function: Callable[..., Any]) -> Callable[..., Any]:
     """
 
     @wraps(noise_scaling_function)
-    def new_scaling_function(circuit: QPROGRAM, *args: Any, **kwargs: Any) -> QPROGRAM:
+    def new_scaling_function(
+        circuit: QPROGRAM, *args: Any, **kwargs: Any
+    ) -> QPROGRAM:
         mitiq_circuit, input_circuit_type = convert_to_mitiq(circuit)
         if kwargs.get("return_mitiq") is True:
             return noise_scaling_function(mitiq_circuit, *args, **kwargs)

--- a/mitiq/conversions.py
+++ b/mitiq/conversions.py
@@ -15,7 +15,7 @@
 
 """Functions for converting to/from Mitiq's internal circuit representation."""
 from functools import wraps
-from typing import Callable, Tuple
+from typing import Any, Callable, Tuple
 
 from cirq import Circuit
 
@@ -58,7 +58,7 @@ def convert_to_mitiq(circuit: QPROGRAM) -> Tuple[Circuit, str]:
     elif isinstance(circuit, Circuit):
         input_circuit_type = "cirq"
 
-        def conversion_function(circ):
+        def conversion_function(circ: Circuit) -> Circuit:
             return circ
 
     else:
@@ -87,6 +87,7 @@ def convert_from_mitiq(circuit: Circuit, conversion_type: str) -> QPROGRAM:
         circuit: Mitiq circuit to convert.
         conversion_type: String specifier for the converted circuit type.
     """
+    conversion_function: Callable[[Circuit], QPROGRAM]
     if conversion_type == "qiskit":
         from mitiq.mitiq_qiskit.conversions import to_qiskit
 
@@ -97,7 +98,7 @@ def convert_from_mitiq(circuit: Circuit, conversion_type: str) -> QPROGRAM:
         conversion_function = to_pyquil
     elif isinstance(circuit, Circuit):
 
-        def conversion_function(circ):
+        def conversion_function(circ: Circuit) -> Circuit:
             return circ
 
     else:
@@ -117,7 +118,7 @@ def convert_from_mitiq(circuit: Circuit, conversion_type: str) -> QPROGRAM:
     return converted_circuit
 
 
-def converter(noise_scaling_function: Callable) -> Callable:
+def converter(noise_scaling_function: Callable[..., Any]) -> Callable[..., Any]:
     """Decorator for handling conversions.
 
     Args:
@@ -126,7 +127,7 @@ def converter(noise_scaling_function: Callable) -> Callable:
     """
 
     @wraps(noise_scaling_function)
-    def new_scaling_function(circuit: QPROGRAM, *args, **kwargs) -> QPROGRAM:
+    def new_scaling_function(circuit: QPROGRAM, *args: Any, **kwargs: Any) -> QPROGRAM:
         mitiq_circuit, input_circuit_type = convert_to_mitiq(circuit)
         if kwargs.get("return_mitiq") is True:
             return noise_scaling_function(mitiq_circuit, *args, **kwargs)

--- a/mitiq/mitiq_pyquil/compiler.py
+++ b/mitiq/mitiq_pyquil/compiler.py
@@ -216,7 +216,7 @@ def is_magic_angle(angle: float) -> bool:
     """
     Checks to see if an angle is 0, +/-pi/2, or +/-pi.
     """
-    return (
+    return bool(
         np.isclose(np.abs(angle), pi / 2)
         or np.isclose(np.abs(angle), pi)
         or np.isclose(angle, 0.0)

--- a/mitiq/mitiq_pyquil/pyquil_utils.py
+++ b/mitiq/mitiq_pyquil/pyquil_utils.py
@@ -13,7 +13,9 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-"""pyQuil utitility functions."""
+"""pyQuil utility functions."""
+from typing import Optional
+
 import numpy as np
 from pyquil import Program
 
@@ -30,7 +32,7 @@ QVM.qam.random_seed = 1337
 np.random.seed(1001)
 
 
-def random_identity_circuit(depth=None):
+def random_identity_circuit(depth: int) -> Program:
     """Returns a single-qubit identity circuit based on Pauli gates."""
 
     # initialize a quantum circuit
@@ -102,7 +104,7 @@ def run_with_noise(circuit: Program, noise: float, shots: int) -> float:
     # we want to simulate noise, so we run without compiling
     results = QVM.run(circuit)
     expval = (results == [0]).sum() / shots
-    return expval
+    return float(expval)
 
 
 def run_program(pq: Program, shots: int = 500) -> float:
@@ -118,7 +120,7 @@ def run_program(pq: Program, shots: int = 500) -> float:
     pq.wrap_in_numshots_loop(shots)
     results = QVM.run(pq)
     expval = (results == [0]).sum() / shots
-    return expval
+    return float(expval)
 
 
 def add_depolarizing_noise(pq: Program, noise: float) -> Program:
@@ -166,7 +168,7 @@ def scale_noise(pq: Program, param: float) -> Program:
     return add_depolarizing_noise(pq, noise)
 
 
-def measure(circuit, qid):
+def measure(circuit: Program, qid: int) -> Program:
     """Returns a circuit adding a register for readout results.
 
     Args:

--- a/mitiq/mitiq_pyquil/pyquil_utils.py
+++ b/mitiq/mitiq_pyquil/pyquil_utils.py
@@ -14,8 +14,6 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 """pyQuil utility functions."""
-from typing import Optional
-
 import numpy as np
 from pyquil import Program
 

--- a/mitiq/mitiq_qiskit/tests/test_conversions_qiskit.py
+++ b/mitiq/mitiq_qiskit/tests/test_conversions_qiskit.py
@@ -115,7 +115,5 @@ def test_convert_with_multiple_barriers(as_qasm):
         cirq_circuit = from_qiskit(qiskit_circuit)
 
     qbit = cirq.LineQubit(0)
-    correct = cirq.Circuit(
-        cirq.ops.H.on(qbit) for _ in range(num_ops)
-    )
+    correct = cirq.Circuit(cirq.ops.H.on(qbit) for _ in range(num_ops))
     assert _equal(cirq_circuit, correct)

--- a/mitiq/utils.py
+++ b/mitiq/utils.py
@@ -15,7 +15,7 @@
 
 """Utility functions."""
 from copy import deepcopy
-from typing import cast
+from typing import cast, Any, Dict
 
 import numpy as np
 
@@ -122,7 +122,7 @@ def _equal(
     )
 
 
-def _are_close_dict(dict_a: dict, dict_b: dict) -> bool:
+def _are_close_dict(dict_a: Dict[Any, Any], dict_b: Dict[Any, Any]) -> bool:
     """Returns True if the two dictionaries have equal keys and
     their corresponding values are "sufficiently" close."""
     keys_a = dict_a.keys()

--- a/mitiq/zne/inference.py
+++ b/mitiq/zne/inference.py
@@ -170,7 +170,9 @@ class Factory(ABC):
         self.opt_params: List[float] = []
         self._already_reduced = False
 
-    def push(self, instack_val: Dict[str, float], outstack_val: float) -> "Factory":
+    def push(
+        self, instack_val: Dict[str, float], outstack_val: float
+    ) -> "Factory":
         """Appends "instack_val" to "self._instack" and "outstack_val" to
         "self._outstack". Each time a new expectation value is computed this
         method should be used to update the internal state of the Factory.

--- a/mitiq/zne/inference.py
+++ b/mitiq/zne/inference.py
@@ -16,7 +16,7 @@
 """Classes corresponding to different zero-noise extrapolation methods."""
 from abc import ABC, abstractmethod
 from copy import deepcopy
-from typing import Callable, Dict, List, Optional, Sequence, Tuple, Union
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union
 import warnings
 
 import numpy as np
@@ -170,7 +170,7 @@ class Factory(ABC):
         self.opt_params: List[float] = []
         self._already_reduced = False
 
-    def push(self, instack_val: dict, outstack_val: float) -> "Factory":
+    def push(self, instack_val: Dict[str, float], outstack_val: float) -> "Factory":
         """Appends "instack_val" to "self._instack" and "outstack_val" to
         "self._outstack". Each time a new expectation value is computed this
         method should be used to update the internal state of the Factory.
@@ -291,7 +291,7 @@ class Factory(ABC):
             max_iterations: Maximum number of iterations (optional).
         """
 
-        def _noise_to_expval(scale_factor, **exec_params) -> float:
+        def _noise_to_expval(scale_factor: float, **exec_params: Any) -> float:
             """Evaluates the quantum expectation value for a given
             scale_factor and other executor parameters."""
             expectation_values = []
@@ -302,7 +302,9 @@ class Factory(ABC):
 
         return self.iterate(_noise_to_expval, max_iterations)
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Factory):
+            return False
         if self._already_reduced != other._already_reduced:
             return False
         if len(self._instack) != len(other._instack):
@@ -397,7 +399,7 @@ class BatchedFactory(Factory):
             )
         return len(self._outstack) == len(self._scale_factors)
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
         return Factory.__eq__(self, other) and np.allclose(
             self._scale_factors, other._scale_factors
         )
@@ -490,7 +492,7 @@ class PolyFactory(BatchedFactory):
         self._already_reduced = True
         return zero_lim
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
         return BatchedFactory.__eq__(self, other) and self.order == other.order
 
 
@@ -750,7 +752,9 @@ class ExpFactory(BatchedFactory):
         self._already_reduced = True
         return zero_lim
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, ExpFactory):
+            return False
         if (
             self.asymptote
             and other.asymptote is None
@@ -890,13 +894,13 @@ class PolyExpFactory(BatchedFactory):
         slope, _ = mitiq_polyfit(scale_factors, exp_values, deg=1)
         sign = np.sign(-slope)
 
-        def _ansatz_unknown(x: float, *coeffs: float):
+        def _ansatz_unknown(x: float, *coeffs: float) -> float:
             """Ansatz of generic order with unknown asymptote."""
             # Coefficients of the polynomial to be exponentiated
             z_coeffs = coeffs[2:][::-1]
             return coeffs[0] + coeffs[1] * np.exp(x * np.polyval(z_coeffs, x))
 
-        def _ansatz_known(x: float, *coeffs: float):
+        def _ansatz_known(x: float, *coeffs: float) -> float:
             """Ansatz of generic order with known asymptote."""
             # Coefficients of the polynomial to be exponentiated
             z_coeffs = coeffs[1:][::-1]
@@ -946,9 +950,10 @@ class PolyExpFactory(BatchedFactory):
         opt_params = [asymptote] + list(z_coefficients[::-1])
         return (zero_lim, opt_params) if full_output else zero_lim
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
         return (
             BatchedFactory.__eq__(self, other)
+            and isinstance(other, PolyExpFactory)
             and np.isclose(self.asymptote, other.asymptote)
             and self.avoid_log == other.avoid_log
             and self.order == other.order
@@ -1182,9 +1187,10 @@ class AdaExpFactory(Factory):
         self._already_reduced = True
         return zero_limit
 
-    def __eq__(self, other) -> bool:
+    def __eq__(self, other: Any) -> bool:
         return (
             Factory.__eq__(self, other)
+            and isinstance(other, AdaExpFactory)
             and self._steps == other._steps
             and self._scale_factor == other._scale_factor
             and np.isclose(self.asymptote, other.asymptote)

--- a/mitiq/zne/scaling/folding.py
+++ b/mitiq/zne/scaling/folding.py
@@ -230,7 +230,7 @@ def _fold_moments(circuit: Circuit, moment_indices: List[int]) -> None:
 
 
 # Helper functions for folding by fidelity
-def _default_weight(op: ops.Operation):
+def _default_weight(op: ops.Operation) -> float:
     """Returns a default weight for an operation."""
     return 0.99 ** len(op.qubits)
 
@@ -291,7 +291,7 @@ def _get_num_to_fold(scale_factor: float, ngates: int) -> int:
 # Local folding functions
 @converter
 def fold_gates_from_left(
-    circuit: QPROGRAM, scale_factor: float, **kwargs
+    circuit: QPROGRAM, scale_factor: float, **kwargs: Any
 ) -> QPROGRAM:
     """Returns a new folded circuit by applying the map G -> G G^dag G to a
     subset of gates of the input circuit, starting with gates at the
@@ -412,7 +412,7 @@ def fold_gates_from_left(
 
 @converter
 def fold_gates_from_right(
-    circuit: QPROGRAM, scale_factor: float, **kwargs
+    circuit: QPROGRAM, scale_factor: float, **kwargs: Any
 ) -> Circuit:
     """Returns a new folded circuit by applying the map G -> G G^dag G
     to a subset of gates of the input circuit, starting with gates at
@@ -489,8 +489,8 @@ def fold_gates_from_right(
 
 
 def _update_moment_indices(
-    moment_indices: dict, moment_index_where_gate_was_folded: int
-) -> dict:
+    moment_indices: Dict[int, int], moment_index_where_gate_was_folded: int
+) -> Dict[int, int]:
     """Updates moment indices to keep track of an original circuit
     throughout folding.
 
@@ -529,7 +529,7 @@ def fold_gates_at_random(
     circuit: QPROGRAM,
     scale_factor: float,
     seed: Optional[int] = None,
-    **kwargs,
+    **kwargs: Any,
 ) -> QPROGRAM:
     """Returns a folded circuit by applying the map G -> G G^dag G to a random
     subset of gates in the input circuit.
@@ -688,7 +688,7 @@ def _fold_local(
     scale_factor: float,
     fold_method: Callable[..., Circuit],
     fold_method_args: Optional[Tuple[Any]] = None,
-    **kwargs,
+    **kwargs: Any,
 ) -> Circuit:
     """Helper function for implementing a local folding method (which nominally
     requires 1 <= scale_factor <= 3) at any scale_factor >= 1. Returns a folded
@@ -750,7 +750,7 @@ def _fold_local(
 
 # Global folding function
 @converter
-def fold_global(circuit: QPROGRAM, scale_factor: float, **kwargs) -> QPROGRAM:
+def fold_global(circuit: QPROGRAM, scale_factor: float, **kwargs: Any) -> QPROGRAM:
     """Returns a new circuit obtained by folding the global unitary of the
     input circuit.
 

--- a/mitiq/zne/scaling/folding.py
+++ b/mitiq/zne/scaling/folding.py
@@ -750,7 +750,9 @@ def _fold_local(
 
 # Global folding function
 @converter
-def fold_global(circuit: QPROGRAM, scale_factor: float, **kwargs: Any) -> QPROGRAM:
+def fold_global(
+    circuit: QPROGRAM, scale_factor: float, **kwargs: Any
+) -> QPROGRAM:
     """Returns a new circuit obtained by folding the global unitary of the
     input circuit.
 

--- a/mitiq/zne/scaling/parameter.py
+++ b/mitiq/zne/scaling/parameter.py
@@ -13,7 +13,7 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-from typing import cast, List, Optional
+from typing import cast, Optional
 import numpy as np
 
 import copy

--- a/mitiq/zne/scaling/parameter.py
+++ b/mitiq/zne/scaling/parameter.py
@@ -13,12 +13,12 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-from typing import Optional
+from typing import cast, List, Optional
 import numpy as np
 
 import copy
 
-from cirq import Circuit, Moment
+from cirq import Circuit, EigenGate, Moment
 from cirq import (
     ZPowGate,
     YPowGate,
@@ -38,10 +38,10 @@ class GateTypeException(Exception):
     pass
 
 
-def _get_base_gate(gate):
+def _get_base_gate(gate: EigenGate) -> EigenGate:
     for base_gate in BASE_GATES:
         if isinstance(gate, base_gate):
-            return base_gate
+            return cast(EigenGate, base_gate)
     raise GateTypeException(
         "Must have circuit be made of rotation gates. "
         "Your gate {} may not be supported".format(gate)
@@ -81,6 +81,7 @@ def scale_parameters(
             if isinstance(gate, MeasurementGate):
                 curr_moment.append(gate(*qubits))
             else:
+                assert isinstance(gate, EigenGate)
                 base_gate = _get_base_gate(gate)
                 param = gate.exponent * np.pi
                 error = rng.normal(loc=0.0, scale=np.sqrt(noise))

--- a/mitiq/zne/zne.py
+++ b/mitiq/zne/zne.py
@@ -14,7 +14,7 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 """High-level zero-noise extrapolation tools."""
-from typing import Callable
+from typing import Callable, Optional
 from functools import wraps
 
 from mitiq._typing import QPROGRAM
@@ -25,7 +25,7 @@ from mitiq.zne.scaling import fold_gates_at_random
 def execute_with_zne(
     qp: QPROGRAM,
     executor: Callable[[QPROGRAM], float],
-    factory: Factory = None,
+    factory: Optional[Factory] = None,
     scale_noise: Callable[[QPROGRAM, float], QPROGRAM] = fold_gates_at_random,
     num_to_average: int = 1,
 ) -> float:
@@ -64,7 +64,7 @@ def execute_with_zne(
 
 def mitigate_executor(
     executor: Callable[[QPROGRAM], float],
-    factory: Factory = None,
+    factory: Optional[Factory] = None,
     scale_noise: Callable[[QPROGRAM, float], QPROGRAM] = fold_gates_at_random,
     num_to_average: int = 1,
 ) -> Callable[[QPROGRAM], float]:
@@ -94,7 +94,7 @@ def mitigate_executor(
 
 
 def zne_decorator(
-    factory: Factory = None,
+    factory: Optional[Factory] = None,
     scale_noise: Callable[[QPROGRAM, float], QPROGRAM] = fold_gates_at_random,
     num_to_average: int = 1,
 ) -> Callable[[Callable[[QPROGRAM], float]], Callable[[QPROGRAM], float]]:

--- a/mypy.ini
+++ b/mypy.ini
@@ -1,2 +1,17 @@
 [mypy]
 ignore_missing_imports = True
+
+# Enable a subset of strict options
+check_untyped_defs = True
+disallow_any_generics = True
+disallow_incomplete_defs = True
+disallow_subclassing_any = True
+# disallow_untyped_calls = True
+disallow_untyped_decorators = True
+disallow_untyped_defs = True
+no_implicit_optional = True
+# no_implicit_reexport = True
+warn_redundant_casts = True
+# warn_return_any = True
+warn_unused_configs = True
+# warn_unused_ignores = True

